spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Linear.scala

```diff
@@ -58,26 +58,24 @@ class Linear(
   override def reset(): Unit = {
     if (initWeight == null) {
       weightInitMethod.init(weight.dense, VariableFormat.OUT_IN)
-      weight.syncToNative()
     } else {
-      weight.copy(initWeight)
+      weight.dense.copy(initWeight)
     }
 
     if (initBias == null) {
       biasInitMethod.init(bias.dense, VariableFormat.ONE_D)
-      bias.syncToNative()
     } else {
-      bias.copy(initBias)
+      bias.dense.copy(initBias)
     }
-
-    gradWeight.zero()
-    gradBias.zero()
   }
 
   override private[mkldnn] def initFwdPrimitives(inputs: Array[MemoryData], phase: Phase) = {
-    val weightShape = inputs(0).shape.length match {
-      case 4 => Array(weight.size(1)) ++ inputs(0).shape.slice(1, 4)
-      case _ => weight.size()
+    val (weightShape, weightLayout) = inputs(0).shape.length match {
+      case 4 =>
+        (Array(weight.size(1)) ++ inputs(0).shape.slice(1, 4),
+          Memory.Format.oihw)
```

Comment:
what's oihw for

Reply:
Linear (mkldnn version) will support 4-D input, so the relevant weight will be 4-D. We make all 4-D weight with the format oihw.